README.txt
This extension is designed to help define contexts for information. It loads the
context into the generic expression parser (see "expressions").
This is used by the Tesseract project but can be applied independently of it.

Refer to the manual for more information.

ChangeLog
2013-05-27 Francois Suter (Cobweb)  <typo3@cobweb.ch>

	* Verified compatibility with TYPO3 CMS 6.1, resolves #48577
	* Released to TER as version 1.4.0

2013-04-01 Francois Suter (Cobweb)  <typo3@cobweb.ch>

	* Verified compatibility with TYPO3 CMS 6.0, resolves #46834
	* Released to TER as version 1.3.0

2012-07-10 Francois Suter (Cobweb)  <typo3@cobweb.ch>

	* Released to TER as version 1.2.0

2012-06-22 Francois Suter (Cobweb)  <typo3@cobweb.ch>

	* Verified compatibility with TYPO3 4.7, raised requirement to TYPO3 4.5, resolves #38282

2012-01-30 Francois Suter (Cobweb)  <typo3@cobweb.ch>

	* Released version 1.1.0 to TER

2012-01-23 Francois Suter (Cobweb)  <typo3@cobweb.ch>

	* Added local storage of context data and retrieval API, resolves #33384

2012-01-11 Francois Suter (Cobweb)  <typo3@cobweb.ch>

	* Added proper support for multidimensional contexts, resolves #33124
	* Added unit tests for new cleanup method

2010-10-27 Francois Suter (Cobweb)  <typo3@cobweb.ch>

	* First public release

tests/tx_context_Test.php
<?php
/***************************************************************
*  Copyright notice
*
*  (c) 2012 Francois Suter <typo3@cobweb.ch>
*  All rights reserved
*
*  This script is part of the TYPO3 project. The TYPO3 project is
*  free software; you can redistribute it and/or modify
*  it under the terms of the GNU General Public License as published by
*  the Free Software Foundation; either version 2 of the License, or
*  (at your option) any later version.
*
*  The GNU General Public License can be found at
*  http://www.gnu.org/copyleft/gpl.html.
*
*  This script is distributed in the hope that it will be useful,
*  but WITHOUT ANY WARRANTY; without even the implied warranty of
*  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
*  GNU General Public License for more details.
*
*  This copyright notice MUST APPEAR in all copies of the script!
***************************************************************/

/**
 * Testcase for the context parser
 *
 * @author		Francois Suter <typo3@cobweb.ch>
 * @package		TYPO3
 * @subpackage	tx_dataquery
 *
 * $Id$
 */
class tx_context_Test extends tx_phpunit_testcase {

		/**
		 * Provides values to parse and the expected result
		 *
		 * @return array
		 */
		public function valuesProvider() {
			$values = array(
					// Value should be unchanged
				'simple value' => array(
					'value' => 'foo',
					'result' => 'foo'
				),
					// Value should be stripped of the "foo:" prefix
				'value with table' => array(
					'value' => 'foo:bar',
					'result' => 'bar'
				),
					// An array of value should remain an array, with prefixed values cleaned up
				'array of values' => array(
					'value' => array(
						'foo' => 'bar',
						'baz' => array(
							'say' => 'hello:world'
						)
					),
					'result' => array(
						'foo' => 'bar',
						'baz' => array(
							'say' => 'world'
						)
					)
				)
			);
			return $values;
		}

		/**
		 * Test a simple value. It should be unchanged
		 *
		 * @param mixed $value Value to test
		 * @param mixed $result Expected result
		 * @test
		 * @dataProvider valuesProvider
		 */
		public function simpleValue($value, $result) {
			/**
			 * @var tx_context	$parser
			 */
			$parser = t3lib_div::makeInstance('tx_context');
			$actualResult = $parser->cleanUpValues($value);
			$this->assertEquals($result, $actualResult);
		}
}
?>

class.tx_context.php
<?php
/***************************************************************
*  Copyright notice
*
*  (c) 2009-2012 Francois Suter (typo3@cobweb.ch)
*  All rights reserved
*
*  This script is part of the TYPO3 project. The TYPO3 project is
*  free software; you can redistribute it and/or modify
*  it under the terms of the GNU General Public License as published by
*  the Free Software Foundation; either version 2 of the License, or
*  (at your option) any later version.
*
*  The GNU General Public License can be found at
*  http://www.gnu.org/copyleft/gpl.html.
*
*  This script is distributed in the hope that it will be useful,
*  but WITHOUT ANY WARRANTY; without even the implied warranty of
*  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
*  GNU General Public License for more details.
*
*  This copyright notice MUST APPEAR in all copies of the script!
***************************************************************/


/**
 * This class provides the functionality for the tx_cachecleaner_cache module of the lowlevel_cleaner
 *
 * @author		Francois Suter <typo3@cobweb.ch>
 * @package		TYPO3
 * @subpackage	tx_context
 *
 *  $Id$
 */
class tx_context {
	protected $extKey = 'context';	// The extension key

	/**
	 * @var array Local copy of the context information
	 */
	static protected $contextData = array();

	/**
	 * This method responds to the configArrayPostProc hook of tslib_fe
	 * It takes the context information from the template and calls on handlers
	 * to load the data where ever necessary
	 *
	 * @param array $params Single entry array containing the "config" part of the template (not used)
	 * @param tslib_fe $parentObject back-reference to the calling object
	 *
	 * @return	void
	 */
	public function loadContext($params, tslib_fe $parentObject) {
		$context = array();
		$tsKey = 'tx_' . $this->extKey . '.';
			// Check for existing context information
		if (isset($params['config'][$tsKey])) {
			$contextSetup = t3lib_div::removeDotsFromTS($params['config'][$tsKey]);
				// Parse the context to make it into a simple hash table
			if (count($contextSetup) > 0) {
				foreach ($contextSetup as $key => $value) {
					$context[$key] = $this->cleanUpValues($value);
				}

					// Store the context data locally
				self::$contextData = $context;
					// Load the context setup into the expression parser's extra data
				tx_expressions_parser::setExtraData($context);
					// Call additional context storing handlers to make the context setup available
					// where ever else it may be needed
				if (is_array($GLOBALS['TYPO3_CONF_VARS']['EXTCONF'][$this->extKey]['contextStorage'])) {
					foreach ($GLOBALS['TYPO3_CONF_VARS']['EXTCONF'][$this->extKey]['contextStorage'] as $className) {
							/** @var $contextStorage tx_context_ContextStorage */
						$contextStorage = t3lib_div::getUserObj($className);
						if ($contextStorage instanceof tx_context_ContextStorage) {
							$contextStorage->storeContext($context);
						}
					}
				}
			}
		}
	}

	/**
	 * Returns a value from the context data
	 *
	 * Multidimensional keys can be passed separated by |, e.g.
	 * 	$pid = tx_context::getContextValue('foo|pid');
	 *
	 * @static
	 * @param string $key Key to search for in the context array
	 * @return mixed Value found in the context array
	 * @throws OutOfRangeException
	 */
	static public function getContextValue($key) {
		if (empty($key)) {
			throw new OutOfRangeException();
		} else {
			$keyList = t3lib_div::trimExplode('|', $key, TRUE);
			$value = self::$contextData;
			while (count($keyList) > 0) {
				$key = array_shift($keyList);
				if (isset($value[$key])) {
					$value = $value[$key];
				} else {
					throw new OutOfRangeException();
				}
			}
		}
		return $value;
	}

	/**
	 * Cleans up value for contexts
	 *
	 * The values may come with the syntax foo:bar where "foo" is expected to be a table name and "bar" a uid
	 * Only the "bar" part should be kept
	 *
	 * @param $value
	 * @return array
	 */
	public function cleanUpValues($value) {
		$returnValue = $value;
		if (is_array($value)) {
			$returnValue = array();
			foreach ($value as $key => $subValue) {
				$returnValue[$key] = $this->cleanUpValues($subValue);
			}

			// If the value contains a colon (:), it means it has a syntax like:
			//		tablename:uid
			// In this case, keep only the uid part
		} elseif (strpos($value, ':') !== FALSE) {
			$valueParts = t3lib_div::trimExplode(':', $value, TRUE);
			if (isset($valueParts[1])) {
				$returnValue = $valueParts[1];
			}
		}
		return $returnValue;
	}
}



if (defined('TYPO3_MODE') && $TYPO3_CONF_VARS[TYPO3_MODE]['XCLASS']['ext/context/class.tx_context.php']) {
	include_once($TYPO3_CONF_VARS[TYPO3_MODE]['XCLASS']['ext/context/class.tx_context.php']);
}
?>

modfunc1/class.tx_context_modfunc1.php
<?php
/***************************************************************
*  Copyright notice
*
*  (c) 2009 Francois Suter (Cobweb) <typo3@cobweb.ch>
*  All rights reserved
*
*  This script is part of the TYPO3 project. The TYPO3 project is
*  free software; you can redistribute it and/or modify
*  it under the terms of the GNU General Public License as published by
*  the Free Software Foundation; either version 2 of the License, or
*  (at your option) any later version.
*
*  The GNU General Public License can be found at
*  http://www.gnu.org/copyleft/gpl.html.
*
*  This script is distributed in the hope that it will be useful,
*  but WITHOUT ANY WARRANTY; without even the implied warranty of
*  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
*  GNU General Public License for more details.
*
*  This copyright notice MUST APPEAR in all copies of the script!
***************************************************************/
/**
 * [CLASS/FUNCTION INDEX of SCRIPT]
 *
 * Hint: use extdeveval to insert/update function index above.
 */


require_once(PATH_t3lib.'class.t3lib_extobjbase.php');



/**
 * Module extension (addition to function menu) 'Context Oveview' for the 'context' extension.
 *
 * @author	Francois Suter (Cobweb) <typo3@cobweb.ch>
 * @package	TYPO3
 * @subpackage	tx_context
 */
class tx_context_modfunc1 extends t3lib_extobjbase {

					/**
					 * Returns the module menu
					 *
					 * @return	Array with menuitems
					 */
					function modMenu()	{
						global $LANG;

						return Array (
							"tx_context_modfunc1_check" => "",
						);
					}

					/**
					 * Main method of the module
					 *
					 * @return	HTML
					 */
					function main()	{
							// Initializes the module. Done in this function because we may need to re-initialize if data is submitted!
						global $SOBE,$BE_USER,$LANG,$BACK_PATH,$TCA_DESCR,$TCA,$CLIENT,$TYPO3_CONF_VARS;

						$theOutput.=$this->pObj->doc->spacer(5);
						$theOutput.=$this->pObj->doc->section($LANG->getLL("title"),"Dummy content here...",0,1);

						$menu=array();
						$menu[]=t3lib_BEfunc::getFuncCheck($this->pObj->id,"SET[tx_context_modfunc1_check]",$this->pObj->MOD_SETTINGS["tx_context_modfunc1_check"]).$LANG->getLL("checklabel");
						$theOutput.=$this->pObj->doc->spacer(5);
						$theOutput.=$this->pObj->doc->section("Menu",implode(" - ",$menu),0,1);

						return $theOutput;
					}
				}



if (defined('TYPO3_MODE') && $TYPO3_CONF_VARS[TYPO3_MODE]['XCLASS']['ext/context/modfunc1/class.tx_context_modfunc1.php'])	{
	include_once($TYPO3_CONF_VARS[TYPO3_MODE]['XCLASS']['ext/context/modfunc1/class.tx_context_modfunc1.php']);
}

?>

interfaces/interface.tx_context_contextstorage.php
<?php
/***************************************************************
*  Copyright notice
*  
*  (c) 2009 Francois Suter (typo3@cobweb.ch)
*  All rights reserved
*
*  This script is part of the TYPO3 project. The TYPO3 project is 
*  free software; you can redistribute it and/or modify
*  it under the terms of the GNU General Public License as published by
*  the Free Software Foundation; either version 2 of the License, or
*  (at your option) any later version.
* 
*  The GNU General Public License can be found at
*  http://www.gnu.org/copyleft/gpl.html.
* 
*  This script is distributed in the hope that it will be useful,
*  but WITHOUT ANY WARRANTY; without even the implied warranty of
*  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
*  GNU General Public License for more details.
*
*  This copyright notice MUST APPEAR in all copies of the script!
***************************************************************/

/** 
 * Interface for the contextStorage hook of class tx_context
 *
 * @author		Francois Suter <typo3@cobweb.ch>
 * @package		TYPO3
 * @subpackage	tx_context
 *
 *  $Id$
 */
interface tx_context_ContextStorage {

	/**
	 * This method receives an array containing the context (simply a hash table)
	 * and should store it according to its needs
	 *
	 * @param	array		$context: List of key-value pairs
	 *
	 * @return	void
	 */
	public function storeContext($context);
}
?>

ext_emconf.php
<?php

/*********************************************************************
* Extension configuration file for ext "context".
*
* Generated by ext 27-05-2013 21:06 UTC
*
* https://github.com/t3elmar/Ext
*********************************************************************/

$EM_CONF[$_EXTKEY] = array (
  'title' => 'Context Manager',
  'description' => 'Provides a general framework to define contexts for parts of the page tree. Other extensions may use these contexts as needed.',
  'category' => 'fe',
  'author' => 'Francois Suter (Cobweb)',
  'author_email' => 'typo3@cobweb.ch',
  'shy' => '',
  'dependencies' => 'expressions',
  'conflicts' => '',
  'priority' => '',
  'module' => '',
  'state' => 'stable',
  'internal' => '',
  'uploadfolder' => 0,
  'createDirs' => '',
  'modify_tables' => '',
  'clearCacheOnLoad' => 0,
  'lockType' => '',
  'author_company' => '',
  'version' => '1.4.0',
  'constraints' => 
  array (
    'depends' => 
    array (
      'typo3' => '4.5.0-6.1.99',
      'expressions' => '',
    ),
    'conflicts' => 
    array (
    ),
    'suggests' => 
    array (
    ),
  ),
  '_md5_values_when_last_written' => 'a:12:{s:9:"ChangeLog";s:4:"5786";s:20:"class.tx_context.php";s:4:"ffab";s:16:"ext_autoload.php";s:4:"2f54";s:12:"ext_icon.gif";s:4:"b6d8";s:17:"ext_localconf.php";s:4:"0bf1";s:16:"locallang_db.xml";s:4:"0445";s:10:"README.txt";s:4:"c12b";s:14:"doc/manual.pdf";s:4:"815d";s:14:"doc/manual.sxw";s:4:"cb36";s:14:"doc/manual.txt";s:4:"ec0e";s:50:"interfaces/interface.tx_context_contextstorage.php";s:4:"0c4a";s:25:"tests/tx_context_Test.php";s:4:"1810";}',
  'suggests' => 
  array (
  ),
  'comment' => 'Verified compatibility with TYPO3 CMS 6.1',
  'user' => 'francois',
);

?>